crates/host/src/wasmbus/mod.rs

```diff
@@ -0,0 +1,642 @@
+                .context("failed to construct key pair from seed")?;
+            ensure!(kp.key_pair_type() == KeyPairType::Cluster);
+            kp
+        } else {
+            KeyPair::new(KeyPairType::Cluster)
+        };
+        let host_key = if let Some(host_seed) = host_seed {
+            let kp =
+                KeyPair::from_seed(&host_seed).context("failed to construct key pair from seed")?;
+            ensure!(kp.key_pair_type() == KeyPairType::Server);
+            kp
+        } else {
+            KeyPair::new(KeyPairType::Server)
+        };
+
+        let labels = HashMap::from([
+            ("hostcore.arch".into(), ARCH.into()),
+            ("hostcore.os".into(), OS.into()),
+            ("hostcore.osfamily".into(), FAMILY.into()),
+            (
+                "path".into(),
+                env::var("WASMCLOUD_HOST_PATH").unwrap_or_default(),
+            ),
+        ]);
+        let friendly_name = names::Generator::default()
+            .next()
+            .context("failed to generate friendly name")?;
+
+        let start_evt = json!({
+            "friendly_name": friendly_name,
+            "labels": labels,
+            "uptime_seconds": 0,
+            "version": env!("CARGO_PKG_VERSION"),
+        });
+
+        let nats = async_nats::connect(url.as_str())
+            .await
+            .context("failed to connect to NATS")?;
+
+        let queue = Queue::new(&nats, &cluster_key, &host_key)
+            .await
+            .context("failed to initialize queue")?;
+        nats.flush().await.context("failed to flush")?;
+
+        let start_at = Instant::now();
+
+        let heartbeat_start_at = start_at
+            .checked_add(Self::HEARTBEAT_INTERVAL)
+            .context("failed to compute heartbeat start time")?;
+        let heartbeat =
+            IntervalStream::new(interval_at(heartbeat_start_at, Self::HEARTBEAT_INTERVAL));
+
+        let (queue_abort, queue_abort_reg) = AbortHandle::new_pair();
+        let (heartbeat_abort, heartbeat_abort_reg) = AbortHandle::new_pair();
+
+        let event_builder = EventBuilderV10::new().source(host_key.public_key());
+        let wasmbus = Lattice {
+            event_builder,
+            friendly_name,
+            host_key,
+            labels,
+            nats,
+            start_at,
+        };
+        wasmbus
+            .publish_event("host_started", start_evt)
+            .await
+            .context("failed to publish start event")?;
+        info!("host {} started", wasmbus.host_key.public_key());
+
+        let wasmbus = Arc::new(wasmbus);
+        let queue = spawn({
+            let wasmbus = Arc::clone(&wasmbus);
+            async {
+                Abortable::new(queue, queue_abort_reg)
+                    .for_each(move |msg| {
+                        let wasmbus = Arc::clone(&wasmbus);
+                        async move { wasmbus.handle_message(msg).await }
+                    })
+                    .await;
```

Comment:
I have not used this syntax before, does this allow this host to handle multiple wasmbus messages simultaneously or is it handled in sequence?

Reply:
This one is sequential, there's https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.for_each_concurrent for concurrent iteration

Reply:
I think we should use the concurrent iteration unless you can see some consequences there, one scenario I can think of is `wadm` sending out 10 commands at the same time and it would be beneficial to handle those concurrently (maybe up to a certain limit) to not delay responses. If it's easy to do now then I think it makes sense to

Reply:
Since this is an optimization, and handling requests out-of-order can have multiple unforeseen side-effects, I think we should wait and do it right instead when we're in position to do so (i.e. the lattice functionality is fully in-place).